cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,527 @@
+This design limits the extensibility of CA and goes against certain native Kubernetes primitives.
+
+This approach aims to make CA pluggable allowing to create experimental providers without needing to change the base CA repository. This new experimental provider only needs to implement the gRPC defined operations.
+
+
+## Appendix
+
+In the following we detail all the message used in the defined actions of the `CloudProvider` GRPC Service server.
+
+### NodeGroups
+
+NodeGroups stores all node groups configured for this cloud provider.
+
+```protobuf
+
+message NodeGroupsRequest {
+  // Intentionally empty.
+}
+
+message NodeGroupsResponse {
+  // All the node groups that the cloud provider service supports. This
+  // field is OPTIONAL.
+  repeated NodeGroup nodeGroups = 1;
+}
+```
+
+### ExternalGrpcNode
+
+ExternalGrpcNode is a custom type. This object defines the minimum required properties of a given Kubernetes node for a node group.
+This new type reduces the amount of data transferred in all the operations instead of
+sending the whole `k8s.io.api.core.v1.Node` rpc message.
+
+```protobuf
+message ExternalGrpcNode{
+	// ID of the node assigned by the cloud provider in the format: <ProviderName>://<ProviderSpecificNodeID>
+	// +optional
+	optional string providerID = 1;
+
+	// Name of the node assigned by the cloud provider
+	optional string name = 2;
+
+	// labels is a map of {key,value} pairs with the node's labels.
+	map<string, string> labels = 3;
+
+
+	// If specified, the node's annotations.
+	map<string, string> annotations = 4;
+}
+```
+
+Initially, we defined a list of 4 properties, but this list could increase during the implementation phase.
+
+### NodeGroupForNode
+
+NodeGroupForNode returns the node group for the given node, nil if the node should not be processed by cluster autoscaler, or non-nil error if such occurred. Must be implemented.
+
+**IMPORTANT:** Please note, this operation is extensively used by CA and can cause some performance degradations on large clusters.
+The initial proposal assumes the rpc responses are cached to offload the performance impact of constantly calling this function.
+
+```protobuf
+message NodeGroupForNodeRequest {
+  // Node group for the given node
+  ExternalGrpcNode node = 1;
+}
+
+message NodeGroupForNodeResponse {
+  // The node group for the given node.
+  repeated NodeGroup nodeGroup = 1;
+}
+
+message NodeGroupForNodeRequest {
+  // Node group for the given node
+  ExternalGrpcNode node = 1;
+}
+
+```
+
+### GetAvailableGPUTypes
+
+GetAvailableGPUTypes handles all available GPU types cloud provider supports.
+
+```protobuf
+message GetAvailableGPUTypesRequest {
+  // Intentionally empty.
+}
+
+message GetAvailableGPUTypesResponse {
+ // GPU types passed in as opaque key-value pairs.
+  map<string, string> gpuTypes = 1;
```

Comment:
actually, in the current interface it returns a map[string]struct{}, making it a string will break changes, imo

Reply:
I changed it now.